### PR TITLE
[WIP] Add `reason` kwarg for message deletion (for audit log support)

### PR DESCRIPTION
**Pull request by krittick**
## Summary

This adds an optional `reason` kwarg to TextChannel.delete_messages, TextChannel.purge, and Message.delete, to support the `X-Audit-Log-Reason` header

## Checklist

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [ ] If `type: ignore` comments were used, a comment is also left explaining why
- [x] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting, examples, ...)

### CODE REVIEW

**Comment by rtk-rnjn** [comment]
Make `reason` keyword only argument

File: discord/channel.py [hunk 1]
```diff
@@ -99,9 +99,9 @@
     from .webhook import Webhook
 
 
-async def _single_delete_strategy(messages: Iterable[Message]):
+async def _single_delete_strategy(messages: Iterable[Message], reason: Optional[str] = None):
```
Comment (BobDotCom) [inline discord/channel.py:102]:
```suggestion
async def _single_delete_strategy(messages: Iterable[Message], *, reason: Optional[str] = None):
```

File: discord/channel.py [hunk 2]
```diff
@@ -376,7 +376,7 @@ async def clone(self, *, name: Optional[str] = None, reason: Optional[str] = Non
             reason=reason,
         )
 
-    async def delete_messages(self, messages: Iterable[Snowflake]) -> None:
+    async def delete_messages(self, messages: Iterable[Snowflake], reason: Optional[str] = None) -> None:
```
Comment (BobDotCom) [inline discord/channel.py:379]:
```suggestion
    async def delete_messages(self, messages: Iterable[Snowflake], *, reason: Optional[str] = None) -> None:
```

**Review by BobDotCom — COMMENTED** [review]
Change `reason` to be keyword-only

File: discord/http.py [hunk 3]
```diff
@@ -252,7 +252,9 @@ async def request(
                 headers["X-Audit-Log-Reason"] = _uriquote(reason, safe="/ ")
 
         kwargs["headers"] = headers
-
+        print(f"For Route: {route.url}")
+        print("Showing headers right before request:")
+        print(kwargs["headers"])
```
Comment (BobDotCom) [inline discord/http.py:257]: Shouldn't these be removed? 
```suggestion
```

Comment (krittick) [reply discord/http.py:257]: That's why it's in draft right now. There are issues on discord's end that I've been trying to narrow down before this is ready for review. The print statements were added during troubleshooting and will be removed once this isn't a draft.

**Comment by krittick** [comment]
This is ready for review.

**Review by Dorukyum — APPROVED** [review]

**Review by Lulalaby — APPROVED** [review]

**Pull request merged**